File: src/main/java/org/cobbzilla/util/reflect/FieldComparator.java
package org.cobbzilla.util.reflect;

import java.util.Comparator;

public class FieldComparator<T, F extends Comparable<F>> implements Comparator<T> {

    private final String field;
    private final boolean reverse = false;

    @java.beans.ConstructorProperties({"field"})
    public FieldComparator(String field) {
        this.field = field;
    }

    @Override public int compare(T o1, T o2) {
        final F v1 = (F) ReflectionUtil.get(o1, field);
        final F v2 = (F) ReflectionUtil.get(o2, field);
        return reverse
                ? (v1 == null ? (v2 == null ? 0 : 1) : (v2 == null ? 1 : v2.compareTo(v1)))
                : (v1 == null ? (v2 == null ? 0 : -1) : (v2 == null ? -1 : v1.compareTo(v2)));
    }

    public String getField() {
        return this.field;
    }

    public boolean isReverse() {
        return this.reverse;
    }
}

File: src/main/java/org/cobbzilla/util/reflect/FieldComparator.kt
package org.cobbzilla.util.reflect

import java.util.Comparator

class FieldComparator<T, F : Comparable<F>> @java.beans.ConstructorProperties("field")
constructor(val field: String) : Comparator<T> {
    val isReverse = false

    override fun compare(o1: T, o2: T): Int {
        val v1 = ReflectionUtil.get(o1, field) as F
        val v2 = ReflectionUtil.get(o2, field) as F
        return if (isReverse)
            if (v1 == null) if (v2 == null) 0 else 1 else v2?.compareTo(v1) ?: 1
        else
            if (v1 == null) if (v2 == null) 0 else -1 else if (v2 == null) -1 else v1.compareTo(v2)
    }
}
